File: pdfsecondmonitor/frmOperation.vb
Public Class frmOperation
#Region "サポートするファイルの種類"

    ''' <summary>
    ''' 開ける動画の拡張子
    ''' </summary>
    Private movieExts = {"avi", "mpeg", "mp4", "wmv", "mov"}

    ''' <summary>
    ''' 開ける画像の拡張子
    ''' </summary>
    Private ImageExts = {"jpeg", "jpg", "bmp", "png", "gif", "tiff", "tif"}

    ''' <summary>
    ''' PDFの拡張子
    ''' </summary>
    Private PDFExts = {"pdf"}
#End Region

#Region "初期処理"

    Private Sub ControlEnable()
        CtlPdf1.ControlEnabled()
        CtlMovie1.ControlEnabled()
        CtlImage1.ControlEnabled()
    End Sub
    Private Sub frmOperation_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        screenDetect()
        AppSettingLoad()
        ActivateForm()
        ControlEnable()
    End Sub

    Private Sub ActivateForm()

        Timer1.Interval = 100
        Timer1.Start()
    End Sub

    Private Sub Timer1_Tick(sender As Object, e As EventArgs) Handles Timer1.Tick
        Me.Activate()
        Timer1.Stop()
    End Sub

#Region "設定値"
    Private loading As Boolean
    Private Sub AppSettingLoad()
        loading = True
        cmbDisplay.SelectedIndex = My.Settings.cmbDisplaySelectedIndex
        Try
            Dim fvinfos As New List(Of FileViewParam)

            'XmlSerializerオブジェクトを作成
            Dim serializer As New System.Xml.Serialization.XmlSerializer(
                GetType(List(Of FileViewParam)))
            '読み込むファイルを開く
            Dim filename = "lstPDFFiles.xml"
            If Not IO.File.Exists(filename) Then
                Return
            End If
            Using sr As New System.IO.StreamReader(
                filename, New System.Text.UTF8Encoding(False))
                'XMLファイルから読み込み、逆シリアル化する
                fvinfos =
                DirectCast(serializer.Deserialize(sr), List(Of FileViewParam))
                'ファイルを閉じる
            End Using
            For Each info In fvinfos
                lstPDFFiles.Items.Add(info)
            Next

        Catch ex As Exception

            MessageBox.Show(ex.Message)
        End Try
        loading = False
    End Sub

    Private Sub AppSettingSave()

        My.Settings.cmbDisplaySelectedIndex = cmbDisplay.SelectedIndex
        My.Settings.Save()

        Dim fvinfos As New List(Of FileViewParam)
        For Each info In lstPDFFiles.Items
            fvinfos.Add(info)
        Next
        Dim serializer As New System.Xml.Serialization.XmlSerializer(
            GetType(List(Of FileViewParam)))
        '書き込むファイルを開く（UTF-8 BOM無し）
        Using sw As New System.IO.StreamWriter(
            "lstPDFFiles.xml", False, New System.Text.UTF8Encoding(False))
            'シリアル化し、XMLファイルに保存する
            serializer.Serialize(sw, fvinfos)
            'ファイルを閉じる
        End Using
    End Sub

#End Region

    Private Sub screenDetect()
        'デバイス名が表示されるようにする
        Me.cmbDisplay.DisplayMember = "DeviceName"
        Me.cmbDisplay.DataSource = Screen.AllScreens

    End Sub
#End Region

#Region "終了処理"

    Private Sub frmOperation_FormClosed(sender As Object, e As FormClosedEventArgs) Handles MyBase.FormClosed
        AppSettingSave()


    End Sub
#End Region

#Region "フォームをセカンドディスプレイに表示"
    Private Sub cmbDisplay_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cmbDisplay.SelectedIndexChanged

        If cmbDisplay.SelectedIndex < 0 Then
            Exit Sub
        End If
        If cmbDisplay.SelectedItem Is Nothing Then
            Exit Sub
        End If
        '        'フォームを表示するディスプレイのScreenを取得する
        Dim s As Screen = DirectCast(Me.cmbDisplay.SelectedItem, Screen)
        '        'フォームの開始位置をディスプレイの左上座標に設定する
        _dispacher.SetSecondScreen(s)




    End Sub



    Private _dispacher As FormDispacher = FormDispacher.GetInstance


#End Region

#Region "リストボックス処理"



    Private Sub btnDelete_Click(sender As Object, e As EventArgs) Handles btnDelete.Click
        If lstPDFFiles.SelectedItem Is Nothing Then
            Exit Sub
        End If
        Dim fileviewinfo As FileViewParam
        fileviewinfo = lstPDFFiles.SelectedItem
        lstPDFFiles.Items.Remove(fileviewinfo)
    End Sub





    Private Sub lstFiles_Click(sender As Object, e As EventArgs) Handles lstPDFFiles.Click
        Dim fileviewinfo As FileViewParam
        fileviewinfo = lstPDFFiles.SelectedItem
        If fileviewinfo Is Nothing Then
            Exit Sub
        End If
        Dim path = fileviewinfo.FileName
        If Not IO.File.Exists(path) Then
            Dim ret = MessageBox.Show("ファイルが見つかりません。リストから削除しますか？", "ファイルがありません", MessageBoxButtons.YesNo)
            If ret = DialogResult.Yes Then
                lstPDFFiles.Items.Remove(fileviewinfo)
            End If
            Exit Sub
        End If
        txtPDFFileName.Text = fileviewinfo.FileName
        Dim ext = IO.Path.GetExtension(fileviewinfo.FileName)
        Dim imageFileViewInfo As FileViewParam = Nothing
        Dim pdfFileViewInfo As FileViewParam = Nothing
        Dim movieFileViewInfo As FileViewParam = Nothing
        If IsPDFExt(ext) Then
            tbcFileOpes.SelectTab(tpAdobePDF.TabIndex)
            pdfFileViewInfo = fileviewinfo
        ElseIf IsImageExt(ext) Then
            tbcFileOpes.SelectTab(tpImage.TabIndex)
            imageFileViewInfo = fileviewinfo
        ElseIf IsMovieExt(ext) Then
            tbcFileOpes.SelectTab(tpMediaPlayer.TabIndex)
            movieFileViewInfo = fileviewinfo
        End If
        CtlImage1.SetFileInfo(imageFileViewInfo)
        CtlMovie1.SetFileInfo(movieFileViewInfo)
        CtlPdf1.SetFileInfo(pdfFileViewInfo)
        ControlEnable()
    End Sub


    Private Function IsContain(ext As String, exts As String()) As Boolean
        For Each target In exts
            If String.Compare($".{target}", ext, True) = 0 Then
                Return True
            End If
        Next
        Return False
    End Function


    Private Function IsPDFExt(ext) As Boolean
        Return IsContain(ext, PDFExts)
    End Function

    Private Function IsImageExt(ext) As Boolean
        Return IsContain(ext, ImageExts)
    End Function

    Private Function IsMovieExt(ext) As Boolean
        Return IsContain(ext, movieExts)
    End Function

    Private Function CreateFilter() As String
        Dim buf As New System.Text.StringBuilder
        buf.Append("画像、動画、PDFファイル")
        buf.Append("|")
        buf.Append(CreateExtsOfFilter(PDFExts))
        buf.Append(CreateExtsOfFilter(ImageExts))
        buf.Append(CreateExtsOfFilter(movieExts))
        buf.Append("|")
        buf.Append("All Files(*.*)")
        buf.Append("|")
        buf.Append("*.*")
        Return buf.ToString()
    End Function

    Private Function CreateExtsOfFilter(ByVal exts As String()) As String
        Dim buf = New System.Text.StringBuilder
        For Each ext In exts
            buf.Append("*.")
            buf.Append(ext)
            buf.Append(";")
        Next
        Return buf.ToString()
    End Function

    Private Sub Button2_Click(sender As Object, e As EventArgs) Handles btnFileAdd.Click
        OpenFileDialog1.Multiselect = True
        OpenFileDialog1.Filter = CreateFilter()
        OpenFileDialog1.FileName = txtPDFFileName.Text
        Dim ret = OpenFileDialog1.ShowDialog()
        If ret = Windows.Forms.DialogResult.Cancel Then
            Exit Sub
        End If

        Dim items = lstPDFFiles.Items

        For Each filename In OpenFileDialog1.FileNames
            items.Add(New FileViewParam(filename))
        Next


    End Sub

    Private Sub btnUnSelect_Click(sender As Object, e As EventArgs) Handles btnUnSelect.Click
        lstPDFFiles.SelectedItem = Nothing
        _dispacher.CloseViewers()
        ControlEnable()
    End Sub

#Region "ドラッグアンドドロップ"
    Private Sub lstPDFFiles_DragEnter(sender As Object, e As DragEventArgs) Handles lstPDFFiles.DragEnter
        'コントロール内にドラッグされたとき実行される
        If e.Data.GetDataPresent(DataFormats.FileDrop) Then
            'ドラッグされたデータ形式を調べ、ファイルのときはコピーとする
            e.Effect = DragDropEffects.Copy
        Else
            'ファイル以外は受け付けない
            e.Effect = DragDropEffects.None
        End If
    End Sub

    Private Sub lstPDFFiles_DragDrop(sender As Object, e As DragEventArgs) Handles lstPDFFiles.DragDrop
        Dim items = lstPDFFiles.Items
        Dim fileName As String() = CType(
                e.Data.GetData(DataFormats.FileDrop, False),
                String())

        For Each f In fileName
            items.Add(New FileViewParam(f))
        Next
    End Sub
#End Region








#End Region

End Class

File: MediaToSecondMonitor/frmOperation.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace MediaToSecondMonitor
{
    public partial class frmOperation : Form
    {
        public frmOperation()
        {
            InitializeComponent();
        }

        /// <summary>
        ///     ''' 開ける動画の拡張子
        ///     ''' </summary>
        private string[] movieExtentions =
        {
        "avi",
        "mpeg",
        "mp4",
        "wmv",
        "mov"
    };

        /// <summary>
        ///     ''' 開ける画像の拡張子
        ///     ''' </summary>
        private string[] ImageExtentions =
        {
        "jpeg",
        "jpg",
        "bmp",
        "png",
        "gif",
        "tiff",
        "tif"
    };

        /// <summary>
        ///     ''' PDFの拡張子
        ///     ''' </summary>
        private string[] PDFExtentions =
        {
        "pdf"
    };


        private void ControlEnable()
        {
            ctlPDF2.ControlEnabled();
            ctlMovie2.ControlEnabled();
            ctlImage2.ControlEnabled();
        }
        private void frmOperation_Load(object sender, EventArgs e)
        {
            screenDetect();
            AppSettingLoad();
            ActivateForm();
            ControlEnable();
        }

        private void ActivateForm()
        {
            Timer1.Interval = 100;
            Timer1.Start();
        }

        private void Timer1_Tick(object sender, EventArgs e)
        {
            this.Activate();
            Timer1.Stop();
        }

        private bool loading;
        private void AppSettingLoad()
        {
            loading = true;
            cmbDisplay.SelectedIndex = Properties.Settings.Default.cmbDisplaySelectedIndex;
            try
            {
                List<FileViewParam> fvinfos = new List<FileViewParam>();

                // XmlSerializerオブジェクトを作成
                System.Xml.Serialization.XmlSerializer serializer = new System.Xml.Serialization.XmlSerializer(typeof(List<FileViewParam>));
                // 読み込むファイルを開く
                var filename = "lstPDFFiles.xml";
                if (!System.IO.File.Exists(filename))
                    return;
                using (System.IO.StreamReader sr = new System.IO.StreamReader(filename, new System.Text.UTF8Encoding(false)))
                {
                    // XMLファイルから読み込み、逆シリアル化する
                    fvinfos = (List<FileViewParam>)serializer.Deserialize(sr);
                }
                foreach (var info in fvinfos)
                    lstPDFFiles.Items.Add(info);
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message);
            }
            loading = false;
        }

        private void AppSettingSave()
        {
            Properties.Settings.Default.cmbDisplaySelectedIndex = cmbDisplay.SelectedIndex;
            Properties.Settings.Default.Save();

            List<FileViewParam> fvinfos = new List<FileViewParam>();
            foreach (FileViewParam info in lstPDFFiles.Items)
                fvinfos.Add(info);
            System.Xml.Serialization.XmlSerializer serializer = new System.Xml.Serialization.XmlSerializer(typeof(List<FileViewParam>));
            // 書き込むファイルを開く（UTF-8 BOM無し）
            using (System.IO.StreamWriter sw = new System.IO.StreamWriter("lstPDFFiles.xml", false, new System.Text.UTF8Encoding(false)))
            {
                // シリアル化し、XMLファイルに保存する
                serializer.Serialize(sw, fvinfos);
            }
        }


        private void screenDetect()
        {
            // デバイス名が表示されるようにする
            this.cmbDisplay.DisplayMember = "DeviceName";
            this.cmbDisplay.DataSource = Screen.AllScreens;
        }


        private void frmOperation_FormClosed(object sender, FormClosedEventArgs e)
        {
            AppSettingSave();
        }

        private void cmbDisplay_SelectedIndexChanged(object sender, EventArgs e)
        {
            if (cmbDisplay.SelectedIndex < 0)
                return;
            if (cmbDisplay.SelectedItem == null)
                return;
            // 'フォームを表示するディスプレイのScreenを取得する
            Screen s = (Screen)this.cmbDisplay.SelectedItem;
            // 'フォームの開始位置をディスプレイの左上座標に設定する
            _dispacher.SecondScreen = s;
        }



        private FormDispacher _dispacher = FormDispacher.Instance;






        private void btnDelete_Click(object sender, EventArgs e)
        {
            if (lstPDFFiles.SelectedItem == null)
                return;
            FileViewParam fileviewinfo
             = (FileViewParam)lstPDFFiles.SelectedItem;
            lstPDFFiles.Items.Remove(fileviewinfo);
        }





        private void lstFiles_Click(object sender, EventArgs e)
        {
            FileViewParam fileviewinfo
            =(FileViewParam)lstPDFFiles.SelectedItem;
            if (fileviewinfo == null)
                return;
            var path = fileviewinfo.FileName;
            if (!System.IO.File.Exists(path))
            {
                var ret = MessageBox.Show("ファイルが見つかりません。リストから削除しますか？", "ファイルがありません", MessageBoxButtons.YesNo);
                if (ret == DialogResult.Yes)
                    lstPDFFiles.Items.Remove(fileviewinfo);
                return;
            }
            txtPDFFileName.Text = fileviewinfo.FileName;
            var ext = System.IO.Path.GetExtension(fileviewinfo.FileName);
            FileViewParam imageFileViewInfo = null/* TODO Change to default(_) if this is not a reference type */;
            FileViewParam pdfFileViewInfo = null/* TODO Change to default(_) if this is not a reference type */;
            FileViewParam movieFileViewInfo = null/* TODO Change to default(_) if this is not a reference type */;
            if (IsPDFExt(ext))
            {
                
                tbcFileOpes.SelectTab(tpAdobePDF.TabIndex);
                pdfFileViewInfo = fileviewinfo;
            }
            else if (IsImageExt(ext))
            {
                tbcFileOpes.SelectTab(tpImage.TabIndex);
                imageFileViewInfo = fileviewinfo;
            }
            else if (IsMovieExt(ext))
            {
                tbcFileOpes.SelectTab(tpMediaPlayer.TabIndex);
                movieFileViewInfo = fileviewinfo;
            }
            ctlImage2.SetFileInfo(imageFileViewInfo);
            ctlMovie2.SetFileInfo(movieFileViewInfo);
            ctlPDF2.FileInfo = pdfFileViewInfo;
            ControlEnable();
        }


        private bool IsContain(string ext, string[] exts)
        {
            foreach (var target in exts)
            {
                if (string.Compare($".{target}", ext, true) == 0)
                    return true;
            }
            return false;
        }


        private bool IsPDFExt(string extention)
        {
            return IsContain(extention, PDFExtentions);
        }

        private bool IsImageExt(string extention)
        {
            return IsContain(extention, ImageExtentions);
        }

        private bool IsMovieExt(string extention)
        {
            return IsContain(extention, movieExtentions);
        }

        private string CreateFilter()
        {
            System.Text.StringBuilder buf = new System.Text.StringBuilder();
            buf.Append("画像、動画、PDFファイル");
            buf.Append("|");
            buf.Append(CreateExtsOfFilter(PDFExtentions));
            buf.Append(CreateExtsOfFilter(ImageExtentions));
            buf.Append(CreateExtsOfFilter(movieExtentions));
            buf.Append("|");
            buf.Append("All Files(*.*)");
            buf.Append("|");
            buf.Append("*.*");
            return buf.ToString();
        }

        private string CreateExtsOfFilter(string[] exts)
        {
            var buf = new System.Text.StringBuilder();
            foreach (var ext in exts)
            {
                buf.Append("*.");
                buf.Append(ext);
                buf.Append(";");
            }
            return buf.ToString();
        }

        private void Button2_Click(object sender, EventArgs e)
        {
            OpenFileDialog1.Multiselect = true;
            OpenFileDialog1.Filter = CreateFilter();
            OpenFileDialog1.FileName = txtPDFFileName.Text;
            var ret = OpenFileDialog1.ShowDialog();
            if (ret == System.Windows.Forms.DialogResult.Cancel)
                return;

            var items = lstPDFFiles.Items;

            foreach (var filename in OpenFileDialog1.FileNames)
                items.Add(new FileViewParam() { FileName = filename });
        }

        private void btnUnSelect_Click(object sender, EventArgs e)
        {
            lstPDFFiles.SelectedItem = null;
            _dispacher.CloseViewers();
            ControlEnable();
        }

        private void lstPDFFiles_DragEnter(object sender, DragEventArgs e)
        {
            // コントロール内にドラッグされたとき実行される
            if (e.Data.GetDataPresent(DataFormats.FileDrop))
                // ドラッグされたデータ形式を調べ、ファイルのときはコピーとする
                e.Effect = DragDropEffects.Copy;
            else
                // ファイル以外は受け付けない
                e.Effect = DragDropEffects.None;
        }

        private void lstPDFFiles_DragDrop(object sender, DragEventArgs e)
        {
            var items = lstPDFFiles.Items;
            string[] fileName = (string[])e.Data.GetData(DataFormats.FileDrop, false);

            foreach (var f in fileName)
                items.Add(new FileViewParam() { FileName = f});
        }
    }
}
